File: go/grundys.go
package main

import (
	"fmt"
	"time"
)

// Computes the Grundy values of Grundy's Game.
func get_grundys(limit int) []int {
	g := make([]int, limit+1)
	gseen := make([]int, limit+1)

	for n := 0; n <= limit; n++ {
		for a := 1; 2*a < n; a++ {
			x := g[a] ^ g[n-a]
			if 2*x < n {
				gseen[x] = n + 1
			}
		}
		mex := 0
		for gseen[mex] == n+1 {
			mex++
		}
		g[n] = mex
	}
	return g
}

// Performance testing
func main() {
	start := time.Now()
	grundys := get_grundys(100_000)
	fmt.Println(grundys[len(grundys)-1]) // 101
	fmt.Println("Computation time (s):", time.Since(start).Seconds())
	// 1.464782 seconds
}

// Use the command "go fmt grundys.go" for formatting

File: c++/grundys.cpp
#include <chrono>
#include <iostream>
#include <vector>

using namespace std;

// Computes the Grundy values of Grundy's Game.
vector<int> grundy_numbers(int limit) {
	vector<int> g(limit + 1, 0);
	vector<int> gseen(limit + 1, 0);

	for (int n = 0; n <= limit; n++) {
		for (int a = 1; 2 * a < n; a++) {
			int x = g[a] ^ g[n - a];
			gseen[x] = n + 1;
		}

		int mex = 0;
		while (gseen[mex] == n + 1) {
			mex++;
		}
		g[n] = mex;
	}

	return g;
}

// Peformance testing
int main() {
	int limit = 100'000;
	auto start = chrono::high_resolution_clock::now();

	vector<int> grundys = grundy_numbers(limit);
	cout << grundys[limit] << endl;
	// 101

	auto end = chrono::high_resolution_clock::now();
	chrono::duration<double> elapsed = end - start;
	cout << "Computation time (s): " << elapsed.count() << endl;
	// 0.430713 seconds
}
